File: 3310-remove-methods-from-project/3310-remove-methods-from-project.py
class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        him=[[] for i in range (n)]
        for a in invocations:
            him[a[0]].append(a[1])
        sp=[False]*n
        self.solve(him,k,sp)
        ans=[]
        got=False
        for a in invocations:
            if not sp[a[0]] and sp[a[1]]:
                got=True
        for i in range(n):
            if(got):
                ans.append(i)
            else:
                if not sp[i]:
                    ans.append(i)
        return ans




    def solve(self,him,k,sp):
        sp[k]=True
        for i in him[k]:
            if not sp[i]:
                self.solve(him,i,sp)

File: 3310-remove-methods-from-project/3310-remove-methods-from-project.java
class Solution {
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        List<List<Integer>> him = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            him.add(new ArrayList<>());
        }
        for (int a[] : invocations) {
            him.get(a[0]).add(a[1]);
        }
        boolean sp[] = new boolean[n];
        solve(sp, him, k);
        List<Integer> ans = new ArrayList<>();
        boolean got = false;
        for (int a[] : invocations) {
            if (!sp[a[0]] && sp[a[1]]) {
                got = true;
                break;
            }
        }

        for (int i = 0; i < n; i++) {
            if (got) {
                ans.add(i);
            } else {
                if (!sp[i]) {
                    ans.add(i);
                }
            }

        }

        return ans;
    }

    void solve(boolean a[], List<List<Integer>> him, int i) {
        a[i] = true;
        
        for (int j : him.get(i)) {
            if (!a[j]) {
                solve(a, him, j);
            }
        }
    }
}